File: Project1/resource/lua_old/game/unit_path.lua
-- Unit Path


function createPath()
	local path = {};
	path.path = createList();
	path.vpath = createList();
	path.pvd = createList();
	path.complete = true;
	
	function path.addTile( tile )
		path.path.add( tile );
		path.vpath.add( tile.address );
		path.pvd.add( tile.pathValDir );
	end
	
	function path.addTileRemote( tile, pvd )
		path.path.add( tile );
		path.vpath.add( tile.address );
		path.pvd.add( pvd );
	end
	
	function path.length()
		return path.path.length();
	end
	
	return path;
end


function createUnitPath( unit, path )
	local unitPath = {};
	unitPath.unit = unit;
	unitPath.path = path;
	unitPath.progressParam = 0.0;
	unitPath.currentTileIndex = 1;
	unitPath.finished = false;
	
	
	-- initialize unit direction
	local pvd = unitPath.path.pvd.get(1);
	local dir0 = unitPath.unit.orientation;
	if( pvd < 0 ) then unitPath.unit.orientation = dir0;
	else unitPath.unit.orientation = pvd; end

	local dir1 = unitPath.unit.orientation;
	local change = (dir1 - dir0 + 6) % 6;
	local angleChange = math.pi * change / 3.0;
	unit.alignToOrientation();
	
	function unitPath.isFinished()
		return unitPath.finished;
	end
	
	function unitPath.getDestination()
		return unitPath.path.path.last();
	end
	
	function unitPath.getAssociatedAP()
		return unitPath.path.path.length();
	end
	
	function unitPath.update()
		-- make progress
		unitPath.progressParam = unitPath.progressParam + 0.075*0.5;
		
		-- are we ready to move on to the next tile?
		if( unitPath.progressParam >= 1.0 ) then
			unitPath.currentTileIndex = unitPath.currentTileIndex + 1;
			
			-- subtract 1.0 rather than setting it to 0 to avoid slowing down the unit
			unitPath.progressParam = unitPath.progressParam - 1.0;
			
			-- update the direction
			local pvd = unitPath.path.pvd.get( unitPath.currentTileIndex );
			local dir0 = unitPath.unit.orientation;
			if( pvd >= 0 ) then unitPath.unit.orientation = pvd; end
			
			local dir1 = unitPath.unit.orientation;
			local change = (dir1 - dir0 + 6) % 6;
			local angleChange = math.pi * change / 3.0;
			unit.alignToOrientation();
		end
		
		-- are we done?
		if( unitPath.currentTileIndex >= unitPath.path.path.length() ) then
			unitPath.finished = true;
			unitPath.unit.isMoving = false;
			return true;
		end
		
		-- set unit position
		unitPath.advanceUnit();
		return false;
	end
	
	
	function unitPath.computeNewPosition()
		local tile0 = unitPath.path.path.get( unitPath.currentTileIndex );
		local tile1 = unitPath.path.path.get( unitPath.currentTileIndex+1 );
		local p0 = tile0.position;
		local p1 = tile1.position;
		local r = p1.sub( p0 ).mul( unitPath.progressParam );
		local p = p0.add( r );
		unitPath.unit.position = p;
		return p;
	end
	
	function unitPath.advanceUnit()
		Unit_advance( unitPath.unit.userdata, unitPath.computeNewPosition() );
	end
	
	
	
	return unitPath;
end

File: Project1/game/UnitPath.cpp
#include "UnitPath.h"
#include "math/Transform.h"
#include "game/Unit.h"

#include <iostream>

using namespace std;
using namespace glm;

UnitPath::UnitPath( Unit *unit, vector<MapTile*> tiles )
	:unit( unit ), tiles( tiles ), finished( false ), currentTileIndex( 0 ), progressParam( 0.0f )
{
	unit->isMoving = true;


	// update Unit direction
	int pvd = tiles.at( 0 )->pathValDir;
	int dir0 = unit->direction;
	unit->direction = pvd < 0 ? dir0 : pvd;
	int dir1 = unit->direction;

	int change = (dir1 - dir0 + 6) % 6;
	float angleChange = math_util::PI * change / 3.0f;
	mat4 rotation = Transform::getRotationY( -angleChange );

	vec3 currPos = unit->transform.position;
	mat4 revTran = Transform::getTranslate( -currPos );
	unit->transform.mul( revTran );
	unit->transform.mul( rotation );
	mat4 forTran = Transform::getTranslate( currPos );
	unit->transform.mul( forTran );
}

UnitPath::~UnitPath()
{

}


bool UnitPath::update()
{
	// make progress
	progressParam += 0.075f;

	// are we ready to move to the next tile?
	if( progressParam >= 1.0f )
	{
		// set index to next Tile
		++currentTileIndex;

		// subtract 1.0 rather than setting to 0.0 to avoid slowing the unit down
		progressParam -= 1.0f;

		// update Unit direction
		int pvd = tiles.at( currentTileIndex )->pathValDir;
		int dir0 = unit->direction;
		unit->direction = pvd < 0 ? dir0 : pvd;
		int dir1 = unit->direction;

		int change = (dir1 - dir0 + 6) % 6;
		float angleChange = math_util::PI * change / 3.0f;
		mat4 rotation = Transform::getRotationY( -angleChange );

		vec3 currPos = unit->transform.position;
		mat4 revTran = Transform::getTranslate( -currPos );
		unit->transform.mul( revTran );
		unit->transform.mul( rotation );
		mat4 forTran = Transform::getTranslate( currPos );
		unit->transform.mul( forTran );
	}

	// are we done?
	if( currentTileIndex >= tiles.size() - 1 )
	{
		finished = true;
		unit->isMoving = false;
		return true;
	}

	// set unit position and orientation
	MapTile *tile0 = tiles.at( currentTileIndex );
	MapTile *tile1 = tiles.at( currentTileIndex + 1 );
	vec3 p0 = tiles.at( currentTileIndex )->position + vec3( 0, tile0->height, 0 );
	vec3 p1 = tiles.at( currentTileIndex + 1 )->position + vec3( 0, tile1->height, 0 );
	vec3 r = (p1 - p0) * progressParam;
	vec3 p = p0 + r;
	
	vec3 currPos = unit->transform.position;
	mat4 revTran = Transform::getTranslate( -currPos );
	unit->transform.mul( revTran );
	mat4 forTran = Transform::getTranslate( p );
	unit->transform.mul( forTran );


	return false;
}
